statsmodels/treatment/panel_effect.py

```diff
@@ -0,0 +1,337 @@
+            return np.linalg.norm(ex.dot(params) - en)
+
+        # constraints and bounds
+        cons = {'type': 'eq', 'fun': lambda x:  np.sum(x)-1}
+        bounds = [[0., 1]] * exog_.shape[1]
+
+        params_start = params0.copy() + 0.01
+        params_start /= params0.sum()
+
+        minout = optimize.minimize(fn, params_start, args=(endog_, exog_),
+                                   method='SLSQP', bounds=bounds,
+                                   constraints=cons)
+        params = minout.x
+
+        return Results(self, params, method='OLSSimplexConstrained')
+
+
+    def predict(self, params, exog):
+        if exog is None:
+            exog = self.exog
+        predicted = exog.dot(params)
+        return predicted
+
+
+class PanelATTBasic(object):
+    """Basic class for treatment effect estimation in panel data
+
+    This is a simple version to get started.
+    Assumes only one treated unit and several controls or predictors.
+
+    Parameters
+    ----------
+    endog : array_like
+        outcome series of the treated unit
+    exog : array_like
+        outcome variable of the control or untreated units,
+        observation in rows, and control units in columns
+    treatment_index : int
+        index of the first treatment period.
+        The average treatment effect is computed using all periods from
+        treatment_index to the end.
+        Estimation of the prediction model uses observations up to the
+        treatment_index (excluding the treatment_index period)
+
+    Notes
+    -----
+    Currently assumes treatment lasts from treatment_index to the end of the
+    series. TODO: add optional end point
+
+    endog, exog is for general prediction
+    We don't separate yet matching variables from prediction, i.e. extra
+    covariates in computing weights as in synthetic control (Abadie et al.)
+
+    This works now in the basic case, but the design has problems to add
+    cross-validation. It might be better to delegate more to the prediction
+    model, e.g. add supporting function for OLS and fit-regularized to do
+    the variable or penalization search.
+
+    Status: experimental, options and API are still changing
+    """
+
+    def __init__(self, endog, exog, treatment_index):
+        self.endog = np.asarray(endog)
+        self.exog = np.asarray(exog)
+        self.treatment_index = treatment_index
+        self.nobs_pre = treatment_index
+        self.slice_pre = slice(None, treatment_index)
+        self.slice_treat = slice(treatment_index, None)
+
+
+    def fit(self, constraints=None, regularization=None, add_const=True):
+        """fit the prediction model
+
+        constraints and regularization are currently mutually exclusive, only
+        one of the too can be different from None.
+
+        I'm adding `add_const` option separately from the constraints because
+        it can be used with OLS and with regularized.
+        (just implmentation leak?)
```

Comment:
spelling.  also can you expand on the implementation leak thought?

Reply:
No idea right now, i.e. I don't remember.
The general point is that we need an option to handle constant separately or to exclude constant shift if we don't want to extrapolate (as imposed in the original articles, but dropped in other versions)